File: src/main/java/com/characohol/characohol_api/controller/RecommendationController.java
package com.characohol.characohol_api.controller;

import com.characohol.characohol_api.domain.Recommendation;
import com.characohol.characohol_api.service.RecommendationService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

@RestController
@RequestMapping("/api/recommendations")
@RequiredArgsConstructor
public class RecommendationController {

    private final RecommendationService recommendationService;

    @GetMapping("/{tag}")
    public Mono<ResponseEntity<Recommendation>> getByTag(@PathVariable String tag) {
        return Mono.fromCallable(() -> recommendationService.getByTag(tag))
                .map(ResponseEntity::ok)
                .onErrorResume(
                        ResponseStatusException.class,
                        e -> Mono.just(ResponseEntity.status(e.getStatusCode()).build())
                )
                .subscribeOn(Schedulers.boundedElastic());
    }

    @GetMapping
    public Flux<Recommendation> getAll() {
        return Flux.defer(() -> Flux.fromIterable(recommendationService.getRecommendations()))
                .subscribeOn(Schedulers.boundedElastic());
    }

    @PostMapping
    public Mono<ResponseEntity<Recommendation>> create(@RequestBody Recommendation req) {
        return Mono.fromCallable(() -> recommendationService.create(req))
                .map(saved -> ResponseEntity.status(HttpStatus.CREATED).body(saved))
                .subscribeOn(Schedulers.boundedElastic());
    }

    @PutMapping("/{id}")
    public Mono<ResponseEntity<Recommendation>> update(@PathVariable Long id, @RequestBody Recommendation req) {
        return Mono.fromCallable(() -> recommendationService.update(id, req))
                .map(ResponseEntity::ok)
                .onErrorResume(ResponseStatusException.class,
                        e -> Mono.just(ResponseEntity.status(e.getStatusCode()).build()))
                .subscribeOn(Schedulers.boundedElastic());
    }

    public Mono<ResponseEntity<Void>> delete(@PathVariable Long id) {
        return Mono.fromRunnable(() -> recommendationService.delete(id))
                .then(Mono.just(ResponseEntity.noContent().<Void>build()))
                .onErrorResume(ResponseStatusException.class,
                        e -> Mono.just(ResponseEntity.status(e.getStatusCode()).build()))
                .subscribeOn(Schedulers.boundedElastic());
    }
}

File: src/main/kotlin/com/characohol/characohol_api/controller/RecommendationController.kt
package com.characohol.characohol_api.controller

import com.characohol.characohol_api.domain.Recommendation
import com.characohol.characohol_api.service.RecommendationService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@RestController
@RequestMapping("/api/recommendations")
class RecommendationController(private val recommendationService: RecommendationService) {

    @GetMapping("/{tag}")
    fun getByTag(@PathVariable tag: String): Mono<ResponseEntity<Recommendation>> {
        return recommendationService.getByTag(tag)
            .map { ResponseEntity.ok(it) }
            .onErrorResume { e ->
                if (e is org.springframework.web.server.ResponseStatusException) {
                    Mono.just(ResponseEntity.status(e.statusCode).build())
                } else {
                    Mono.just(ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build())
                }
            }
    }

    @GetMapping
    fun getAll(): Flux<Recommendation> {
        return recommendationService.getRecommendations()
    }

    @PostMapping
    fun create(@RequestBody req: Recommendation): Mono<ResponseEntity<Recommendation>> {
        return recommendationService.create(req)
            .map { ResponseEntity.status(HttpStatus.CREATED).body(it) }
    }

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long, @RequestBody req: Recommendation): Mono<ResponseEntity<Recommendation>> {
        return recommendationService.update(id, req)
            .map { ResponseEntity.ok(it) }
            .onErrorResume { e ->
                if (e is org.springframework.web.server.ResponseStatusException) {
                    Mono.just(ResponseEntity.status(e.statusCode).build())
                } else {
                    Mono.just(ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build())
                }
            }
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long): Mono<ResponseEntity<Void>> {
        return recommendationService.delete(id)
            .then(Mono.just(ResponseEntity.noContent().build<Void>()))
            .onErrorResume { e ->
                if (e is org.springframework.web.server.ResponseStatusException) {
                    Mono.just(ResponseEntity.status(e.statusCode).build())
                } else {
                    Mono.just(ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build())
                }
            }
    }
}